### PR TITLE
Revise and optimize the RPC method `trace_callMany`

### LINKED ISSUE #4165: trace_callMany doesn't allow to trace dependent transactions
**Describe the bug**
While not being a standard ETH RPC method, `trace_callMany` was first introduced by the OE in order to allow simulating dependent transactions (cf. [docs](https://openethereum.github.io/JSONRPC-trace-module#trace_callmany)). It has since been implemented by other clients such as Erigon (https://github.com/ledgerwatch/erigon/pull/1486) and also Nethermind (in #3571).

However, Nethermind's implementation differs from the other clients in that it doesn't apply state changes in between calls, thus if call1 updates state (e.g. via a transfer), call2 doesn't have access to the updated state (e.g. the new balance).

This makes it hard for us to use Nethermind as a drop in replacement for other nodes, as our offchain infrastructure uses `trace_callMany` to simulate dependent calls.

**To Reproduce**
E.g. the following `trace_callMany` (send tokens from acc1 to acc2, then sending tokens from acc2 back to acc1) fails in the second transaction:

```
curl <your URL here>  \
  -X POST \
  -H "Content-Type: application/json" \
  -d "{\"jsonrpc\":\"2.0\",\"method\":\"trace_callMany\",\"params{\"data\":\"0xa9059cbb000000000000000000000000000000000000000000000000000000000000def100000000000000000000000000000000000000000000000000000000000003e8\",\"from\":\"0xca771eda0c70aa7d053ab1b25004559b918fe662\",\"to\":\"0xdef1ca1fb7fbcdc777520aa7f396b4e015f497ab\"},[\"trace\"]],[{\"data\":\"0xa9059cbb000000000000000000000000ca771eda0c70aa7d053ab1b25004559b918fe66200000000000000000000000000000000000000000000000000000000000003e8\",\"from\":\"0x000000000000000000000000000000000000def1\",\"to\":\"0xdef1ca1fb7fbcdc777520aa7f396b4e015f497ab\"},[\"trace\"]]],\"latest\"],\"id\":1}"
```

**Expected behavior**
It should not revert (expect same positive result as on Erigon/OE nodes)

### PR DESCRIPTION
**Pull request by rubo**
Fixes #4165. Closes #4419.

## Changes:

This is a revised and optimized version of PR #4419.
This also changes trace module from arrays to enumerables to avoid unnecessary allocations.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

Besides unit tests, the changes have been tested on Goerli as follows:

1. Transfer funds from account A to an unused account B and then back to account A. Succeded as expected.
2. Transfer funds from account A to account B and then from an unused account C to account A. Failed as expected.

### CODE REVIEW

**Comment by rubo** [comment]
@LukaszRozmej Geth uses `debug_traceCall` naming. I'm not sure how our naming aligns with the Debug module standardization.

**Comment by tanishqjasoria** [comment]
> @LukaszRozmej Geth uses `debug_traceCall` naming. I'm not sure how our naming aligns with the Debug module standardization.

this call is part of the trace module that openethereum introduced: https://openethereum.github.io/JSONRPC-trace-module

**Review by tanishqjasoria — APPROVED** [review]

File: src/Nethermind/Nethermind.Evm/Tracing/ParityStyle/ParityLikeBlockTracer.cs [hunk 1]
```diff
@@ -1,34 +1,32 @@
...
-using Nethermind.Core.Specs;
 using Nethermind.Int256;
 
 namespace Nethermind.Evm.Tracing.ParityStyle
 {
     public class ParityLikeBlockTracer : BlockTracerBase<ParityLikeTxTrace, ParityLikeTxTracer>
     {
+        private readonly Dictionary<Keccak, ParityTraceTypes>? _typesByTransaction;
         private Block _block = null!;
-        private bool _validateChainId;
```
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Evm/Tracing/ParityStyle/ParityLikeBlockTracer.cs:29]: Not sure but wasn't used.

**Review by LukaszRozmej — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
> @LukaszRozmej Geth uses `debug_traceCall` naming. I'm not sure how our naming aligns with the Debug module standardization.

`debug` and `trace` modules are different things and use different formats.

**Comment by LukaszRozmej** [comment]
Could you also smoke test other changed tracing methods (they should be fine, but...)

File: src/Nethermind/Nethermind.Evm/Tracing/ParityStyle/ParityLikeBlockTracer.cs [hunk 2]
```diff
@@ -42,45 +40,43 @@ public ParityLikeBlockTracer(ParityTraceTypes types)
             _types = types;
             IsTracingRewards = (types & ParityTraceTypes.Rewards) == ParityTraceTypes.Rewards;
         }
-        
-        public ParityLikeBlockTracer(ParityTraceTypes types, ISpecProvider specProvider)
+
+        public ParityLikeBlockTracer(Dictionary<Keccak,ParityTraceTypes> typesByTransaction)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Tracing/ParityStyle/ParityLikeBlockTracer.cs:44]: Should be IDictionary

**Pull request merged**